recirq/otoc/parallel_xeb.py

```diff
@@ -0,0 +1,725 @@
+        gate_list_corrected = generic_fsim_gate(corrected_angles, (q_0, q_1))
+        circuit_corrected = cirq.Circuit(gate_list_corrected)
+        correction_gates[(q0, q1)] = circuit_corrected
+
+        fig = plt.figure()
+        plt.plot(
+            num_cycle_range, f_vals_0, 'ro', figure=fig,
+            label=r'{} and {}, unoptimized [$r_p$ = {}]'.format(
+                q0, q1, err_0.__round__(5)))
+        plt.plot(
+            num_cycle_range, f_vals_1, 'bo', figure=fig,
+            label=r'{} and {}, optimized [$r_p$ = {}]'.format(
+                q0, q1, err_1.__round__(5)))
+        plt.plot(
+            num_cycle_range, sqrt_purities, 'go', figure=fig,
+            label=r'{} and {}, purity error = {}'.format(
+                q0, q1, err_p.__round__(5)))
+        plt.plot(x_fitted_0, y_fitted_0, 'r--')
+        plt.plot(x_fitted_1, y_fitted_1, 'b--')
+        plt.plot(x_vals_p, y_vals_p, 'g--')
+        plt.legend()
+        plt.xlabel('Number of Cycles')
+        plt.ylabel(r'XEB Fidelity')
+
+        if save_directory is not None:
+            fig.savefig(save_directory + '/xeb_q{}_{}_q{}_{}'.format(
+                q0[0], q0[1], q1[0], q1[1]))
+
+        if not plot_individual_traces:
+            plt.close(fig)
+
+    num_pairs = len(final_errors_optimized)
+    pair_pos = np.linspace(0, 1, num_pairs)
+
+    fig_0 = plt.figure()
+    plt.plot(sorted(final_errors_unoptimized.values()), pair_pos,
+             figure=fig_0, label='Unoptimized Unitaries')
+    plt.plot(sorted(final_errors_optimized.values()), pair_pos,
+             figure=fig_0, label='Optimized Unitaries')
+    plt.plot(sorted(purity_errors.values()), pair_pos,
+             figure=fig_0, label='Purity Errors')
+    plt.xlabel(r'Pauli Error Rate, $r_p$')
+    plt.ylabel(r'Integrated Histogram')
+    plt.legend()
+
+    fig_1 = plt.figure()
+    for label in _fsim_angle_labels:
+        shifts = [a[label] for a in delta_angles.values()]
+        plt.plot(sorted(shifts), pair_pos, figure=fig_1, label=label)
+    plt.xlabel(r'FSIM Angle Error (Radian)')
+    plt.ylabel(r'Integrated Histogram')
+    plt.legend()
+
+    if not plot_histograms:
+        plt.close(fig_0)
+        plt.close(fig_1)
+
+    if save_directory is not None:
+        fig_0.savefig(save_directory + '/pauli_error_histogram')
+        fig_1.savefig(save_directory + '/angle_shift_histogram')
+
+    return (fitted_gates, correction_gates, fitted_angles,
+            final_errors_optimized, final_errors_unoptimized)
+
+
+def single_qubit_xeb_fidelities(
+        all_qubits: List[Tuple[int, int]],
+        num_cycle_range: Sequence[int],
+        measured_bits: List[List[np.ndarray]],
+        scrambling_gates: List[np.ndarray],
+        plot_individual_traces: bool = False,
+        plot_histograms: bool = False,
+        save_directory: Optional[str] = None
+) -> Dict[Tuple[int, int], float]:
+    """Computes single-qubit gate fidelities from parallel-XEB data.
```

Comment:
how does this work? Is the user supposed to run parallel single-qubit circuits?

Reply:
This is historical. I deleted the single-qubit XEB code. It was occasionally used as a sanity check on the single-qubit gates, but hasn't been particularly useful for the OTOC experiment,